pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+	}
+	if errs != nil {
+		return logAndReturnReconcileErrors(errs)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// Start out with everything in m and only retain what's not in supplied endpoints.
+	epsToDelete := make(map[string]*Endpoint, len(m.allEndpoints))
+	for uid, ep := range m.allEndpoints {
+		epsToDelete[uid] = ep
+	}
+	for uid, expectedEp := range epsToAdd {
+		existingEp, found := epsToDelete[uid]
+		if !found {
+			continue // expectedEp will be added
+		}
+		if !reflect.DeepEqual(*expectedEp, *existingEp) {
+			continue // expectedEp will be added, existingEp will be deleted
+		}
+		// endpoint metadata has remained the same, do not add or delete
+		delete(epsToAdd, uid)
+		delete(epsToDelete, uid)
+	}
+	if len(epsToAdd) == 0 && len(epsToDelete) == 0 {
+		return nil
+	}
+	for uid, delEp := range epsToDelete {
+		m.deleteEndpoint(uid, delEp)
+	}
+	newLabelMappings := attributeValues{}
+	newSubsetMappings := subsetEndpoints{}
+	for uid, addEp := range epsToAdd {
+		newLabelMappings.addServiceEndpoint(uid, addEp)
+		m.allEndpoints[uid] = addEp
+		// By default we create a subset with the same name as the service.
+		// This provides a set of subsets that do not have any bearings on endpoint
+		// labels and is used for CDS.
+		// TODO: No effort is being made to ensure Subset names are distinct from
+		// service names. If this is deemed necessary, one option is to create a unique
+		// service-subsetname key. Alternatively galley can create subsets with the name
+		// of the service and empty labels. In any case, populating these default
+		// subsets may be OK if pilot should not need to wait for configs supplied by
+		// galley, given that controllers may already provide Mesh with this info.
+		svcName, _ := addEp.getLabelValue(SERVICE.stringValue())
+		newSubsetMappings.addEndpointUID(svcName, uid)
+	}
+	// Verify if existing subset definitions apply for the newly added labels
+	for ssName, subset := range m.subsetDefinitions {
+		matchingEpKeys := newLabelMappings.getKeysMatching(subset.Labels)
+		newSubsetMappings.addEndpointUIDs(ssName, matchingEpKeys)
+	}
+	// Update mesh for new label mappings and new subset mappings
+	m.reverseAttrMap.addAttributeValues(newLabelMappings)
+	m.subsetEndpoints.addSubsetEndpoints(newSubsetMappings)
+	return nil
+}
+
+// MeshEndpoints implements functionality required for EDS and returns a list of endpoints
+// that match one or more subsets.
+func (m *Mesh) MeshEndpoints(subsetNames []string) []*Endpoint {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	epSet := endpointUIDSet{}
+	for _, ssName := range subsetNames {
+		ssEpSet, found := m.subsetEndpoints[ssName]
+		if !found {
+			continue
+		}
+		epSet.appendAll(ssEpSet)
+	}
+	out := make([]*Endpoint, len(epSet))
+	idx := 0
+	for uid, _ := range epSet {
+		out[idx] = m.allEndpoints[uid]
+		idx++
+	}
+	return out
+}
+
+// SubsetNames implements functionality required for CDS and returns a list of all subset
+// names currently configured for this Mesh
```

Comment:
Why do we call subsets clusters? If it's the same thing, can that be stated somewhere, since it helps understanding. We are not going to use `https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/cluster/cluster.proto#L241`? It's very confusing to have subsets and cluster subsets at the same time.

Reply:
EDS is keyed by cluster names, not by subset names. There's no field in EDS to pass a list of subsets in clusters.

Reply:
See my final comment. I propose to get rid of this Subset term and fully embrace clusters and endpoints. All we need here is map of clusters, endpoints, and some reverse indices. It is immaterial whether those clusters come from a subset or from the default service.

Reply:
Hmmm ...  using the term "clusters" in the way Envoy uses it ... has known to be confusing with customers and readers of the design doc. Also tying this to Envoy is probably not the best thing .... given that Envoy v1 to v2 API has unnecessarily caused 2 revs for this very PR. We should focus on Istio terms rather than being tripped up by Envoy or Kubernetes terms.